### ISSUE TITLE
Keystone authentication

### ISSUE THREAD

**Issue by jpforny** [EGI-FCTF/jOCCI-api#11]
I'm not being able to authenticate against Keystone and didn't found any usage example for this use case. Normally I would request a token to Keystone endpoint and use it in every request to occi-os (https://github.com/tmetsch/occi-os) endpoint. Since I had no success trying to use KeystoneAuthentication class, I added the token into the client connection headers, but got error 405 in the HEAD request of the authenticate method. Guess I might be missing something or it is some specific problem with occi-os. 

How is the proper way to create a client for this case with jOCCI-api? If you need any more details please let me know.

Thanks!

**Comment by spinto** [comment]
Sono fuori dall'ufficio da Lun 21/12/2015 fino a Lun 11/01/2016.

For urgent matter, you can contact the TEP Core Team (tepcoreteam@esa.int).


Note: This is an automated response to your message  "[jOCCI-api] Keystone
authentication (#11)" sent on 1/7/2016 14:15:18.

This is the only notification you will receive while this person is away.


This message and any attachments are intended for the use of the addressee or addressees only.
The unauthorised disclosure, use, dissemination or copying (either in whole or in part) of its
content is not permitted.
If you received this message in error, please notify the sender and delete it from your system.
Emails can be altered and their integrity cannot be guaranteed by the sender.

Please consider the environment before printing this email.

**Comment by amarthadan** [comment]
Hi. `KeystoneAuthentication` class is used as a fallback for other authentication methods (this is a bit misleading in our Wiki and I will update it). It means that you will choose one of other authentication methods that has `KeystoneAuthentication` as a fallback. You can choose from `BasicAuthentication`, `DigestAuthentication`, `X509Authentication` or `VOMSAuthentication`. Next, you have to set jOCCI client's endpoint to server's OCCI endpoint NOT Keystone endpoint. Server will then response with an URL of Keystone endpoint and jOCCI client is expecting that. jOCCI client will then authenticate against Keystone endpoint with selected authentication method.

So, to sum this up, use authentication method other then `KeystoneAuthentication` against server's OCCI endpoint. Redirect and authentication against Keystone is basically in-built. Also, you can check out our [examples](https://github.com/EGI-FCTF/jOCCI-api/tree/master/src/main/java/cz/cesnet/cloud/occi/api/example).

**Comment by spinto** [comment]
Sono fuori dall'ufficio da Lun 21/12/2015 fino a Lun 11/01/2016.

For urgent matter, you can contact the TEP Core Team (tepcoreteam@esa.int).


Note: This is an automated response to your message  "Re: [jOCCI-api]
Keystone authentication (#11)" sent on 1/8/2016 9:43:23.

This is the only notification you will receive while this person is away.


This message and any attachments are intended for the use of the addressee or addressees only.
The unauthorised disclosure, use, dissemination or copying (either in whole or in part) of its
content is not permitted.
If you received this message in error, please notify the sender and delete it from your system.
Emails can be altered and their integrity cannot be guaranteed by the sender.

Please consider the environment before printing this email.

**Comment by jpforny** [comment]
Hi. Thanks for the reply, it helped me understand the fallback concept. I used `BasicAuthentication`, which retrieved a token from Keystone, but I'm getting error 401 inside `tryTenants` method though.

I'm aware that OpenStack provides a URI (tag Www-Authenticate) in a response with code 401, but I'm only getting another 401 using the token obtained from this URI. I tested via curl the requests to the `MODEL_URI` and only received code 200 using a token obtained from http://host:5000/v3/auth/tokens. I'm not sure what is the best solution or if it is some environment problem. Any suggestion would be really appreciated.

Thank you!

**Comment by amarthadan** [comment]
Keystone v3 has different authentication mechanisms then Keystone v2. jOCCI currently doesn't support Keystone v3. If you want to use jOCCI, you will have to set occi-os to use Keystone v2.

**Comment by jpforny** [comment]
Ok!

Thank you very much for your help!
